### PR TITLE
[build] [ci] Enable code coverage support.

### PR DESCRIPTION
**Pull request by ejgallego**
We add a new travis target that will generate coverage information for
the test suite and upload it to coveralls using bisect_ppx.

We could also generate a regular `bisect_ppx` report.

This depends on #7150

### CODE REVIEW

File: Makefile.build [hunk 1]
```diff
@@ -765,6 +765,21 @@ $(VDFILE).d: $(D_DEPEND_BEFORE_SRC) $(VFILES) $(D_DEPEND_AFTER_SRC) $(COQDEPBOOT
...
 
+###########################################################################
+# Bisect Coverage
+###########################################################################
+.PHONY: bisect_report
+
+bisect_report: coveralls.json
+
+BISECT_FILES=bisect*.out
+# use ocveralls --send ?
```
Comment (SkySkimmer) [inline Makefile.build]: typo

Comment (SkySkimmer) [reply Makefile.build]: ?

File: Makefile.build [hunk 2]
```diff
@@ -765,6 +765,21 @@ $(VDFILE).d: $(D_DEPEND_BEFORE_SRC) $(VFILES) $(D_DEPEND_AFTER_SRC) $(COQDEPBOOT
...
+coveralls.json: OCAMLC := $(OCAMLFIND) ocamlc $(CAMLFLAGS) -package bisect_ppx
+coveralls.json: OCAMLOPT := $(OCAMLFIND) opt $(CAMLFLAGS) -package bisect_ppx
+coveralls.json: test-suite
+	ocveralls $(BISECT_FILES) > coveralls.json
+	bisect-ppx-report -html coverage/ $(BISECT_FILES)
```
Comment (SkySkimmer) [inline Makefile.build]: This line makes `coverage/index.html` (and others) and so is in the wrong target.

File: Makefile.build [hunk 3]
```diff
@@ -765,6 +765,21 @@ $(VDFILE).d: $(D_DEPEND_BEFORE_SRC) $(VFILES) $(D_DEPEND_AFTER_SRC) $(COQDEPBOOT
...
+# Bisect Coverage
+###########################################################################
+.PHONY: bisect_report
+
+bisect_report: coveralls.json
+
+BISECT_FILES=bisect*.out
+# use ocveralls --send ?
+coveralls.json: OCAMLC := $(OCAMLFIND) ocamlc $(CAMLFLAGS) -package bisect_ppx
+coveralls.json: OCAMLOPT := $(OCAMLFIND) opt $(CAMLFLAGS) -package bisect_ppx
```
Comment (SkySkimmer) [inline Makefile.build]: Why not use a configure option? Since it requires a rebuild to change that seems like the appropriate place to do it.

File: Makefile.build [hunk 4]
```diff
@@ -765,6 +765,21 @@ $(VDFILE).d: $(D_DEPEND_BEFORE_SRC) $(VFILES) $(D_DEPEND_AFTER_SRC) $(COQDEPBOOT
 	$(SHOW)'COQDEP    VFILES'
 	$(HIDE)$(COQDEPBOOT) -boot $(DYNDEP) $(VFILES) $(TOTARGET)
 
+###########################################################################
+# Bisect Coverage
+###########################################################################
```
Comment (SkySkimmer) [inline Makefile.build]: Wouldn't this fit `Makefile.dev` better?

**Comment by SkySkimmer** [comment]
Do you intend to make gitlab/circle versions too? Did you try setting up the travis integration https://docs.travis-ci.com/user/coveralls/?
With circle/gitlab it should be possible to aggregate coverage over everything we have in CI which I think would be cool.

File: Makefile.build (hunk 3 above)
Comment (ejgallego) [reply Makefile.build]: I am moving away from configure, which is a big pain, in the future this will be place under a "workspace".

File: Makefile.build (hunk 2 above)
Comment (ejgallego) [reply Makefile.build]: `coverage/index.html` is just an extra.

File: Makefile.build (hunk 3 above)
Comment (ejgallego) [reply Makefile.build]: Note that indeed this kind of package dependency cannot be solved by make or configure in a proper way.

File: Makefile.build (hunk 4 above)
Comment (ejgallego) [reply Makefile.build]: Maybe, but I always get pain with the `test-suite` target not being in scope.

**Comment by ejgallego** [comment]
> Do you intend to make gitlab/circle versions too?

Yes / No.

> Did you try setting up the travis integration https://docs.travis-ci.com/user/coveralls/?

Next step once this builds in Travis.

> With circle/gitlab it should be possible to aggregate coverage over everything we have in CI which I think would be cool.

Indeed that'd be very cool stuff. I don't plan to work a lot more on this thou until Dune support lands in. My goal for this PR is just to get the basic Travis report out and ensure my tree doesn't stagnate.

File: Makefile.build (hunk 2 above)
Comment (SkySkimmer) [reply Makefile.build]: >coverage/index.html is just an extra.

So?

File: Makefile.build (hunk 3 above)
Comment (SkySkimmer) [reply Makefile.build]: >Note that indeed this kind of package dependency cannot be solved by make or configure in a proper way.

make `configure -bisect-ppx` add the flag to CAMLFLAGS, problem solved no?

Comment (ejgallego) [reply Makefile.build]: That seems overkill.

**Comment by ejgallego** [comment]
Yeah maybe you are right and we should modify configure as Gabriel suggested to have extra custom caml byte and opt flags, and deprecate `-flambda-opts` by the way.

File: Makefile.build (hunk 3 above)
Comment (SkySkimmer) [reply Makefile.build]: >That seems overkill.

It seems like the right way to change the compilation.
Also allows me to easily run with bisect-ppx on random developments locally.

File: Makefile.ide [hunk 5]
```diff
@@ -95,8 +95,7 @@ ide-toploop: ide-$(BEST)loop
 ifeq ($(HASCOQIDE),opt)
 $(COQIDE): $(LINKIDEOPT)
 	$(SHOW)'OCAMLOPT -o $@'
-	$(HIDE)$(OCAMLOPT) $(COQIDEFLAGS) $(OPTFLAGS) -o $@ unix.cmxa threads.cmxa lablgtk.cmxa \
-		lablgtksourceview2.cmxa str.cmxa $(IDEFLAGS:.cma=.cmxa) $^
+	$(HIDE)$(OCAMLOPT) $(COQIDEFLAGS) $(OPTFLAGS) -package str,lablgtk2.sourceview2 -linkpkg $(IDEFLAGS:.cma=.cmxa) -o $@ $^
```
Comment (SkySkimmer) [inline Makefile.ide]: Why this change?

Comment (ejgallego) [reply Makefile.ide]: Otherwise linking will fail, as the base modules are compiled with bisect and need the bisect unit. Note that this change is already present on #6859 so things will work "magically" there.

ocamlfind does some magic when using `-linkpkg` so indeed if the `cma` depends on bisect it will be automatically added to close the object dependency chain.

File: Makefile.build (hunk 3 above)
Comment (ejgallego) [reply Makefile.build]: Yup you are right, so once I have a minute I'll do a patch adding `-camlflags -camlbyteflags -camloptflags` to `configure` and rebase this "patch" on top of it. [So indeed there won't be a `-bisect ppx` option, but Travis will just do `-camlflags "-package ppx_bisect"`.

Comment (ejgallego) [reply Makefile.build]: Feel free to take over that configure patch if you want; I am not gonna do it immediately, I just pushed this PR as not to let the tree bitrot on my HD.

Comment (ejgallego) [reply Makefile.build]: Never mind, I am finally doing this now.

**Comment by ejgallego** [comment]
Ok so this is now following a configure option as @SkySkimmer suggested. Still many refinements to do, but let's see how Travis / coveralls fare now.

File: .travis.yml [hunk 6]
```diff
@@ -52,9 +52,10 @@ env:
   # Main test suites
   matrix:
   - TEST_TARGET="test-suite" COMPILER="4.02.3+32bit"
+  - TEST_TARGET="coverage_report" EXTRA_OPAM="bisect_ppx ocveralls" COMPILER="$COMPILER_BE" CAMLP5_VER="$CAMLP5_VER_BE" FINDLIB_VER="$FINDLIB_VER_BE" EXTRA_CONF="-camloptflags '-package bisect_ppx'"
   - TEST_TARGET="validate"                           TW="travis_wait"
   - TEST_TARGET="validate"   COMPILER="4.02.3+32bit" TW="travis_wait"
-  - TEST_TARGET="validate"   COMPILER="${COMPILER_BE}+flambda" CAMLP5_VER="${CAMLP5_VER_BE}" NATIVE_COMP="no" EXTRA_CONF="-flambda-opts -O3" EXTRA_OPAM="num" FINDLIB_VER="${FINDLIB_VER_BE}"
+  - TEST_TARGET="validate"   COMPILER="${COMPILER_BE}+flambda" CAMLP5_VER="${CAMLP5_VER_BE}" NATIVE_COMP="no" EXTRA_CONF="-camloptflags '-O3 -unbox-closures'" EXTRA_OPAM="num" FINDLIB_VER="${FINDLIB_VER_BE}"
```
Comment (SkySkimmer) [inline .travis.yml]: Does this actually work? It seems like it's going to expand in a bad way.
ie configure will see `'-O3` `-unbox-closures'`

Comment (ejgallego) [reply .travis.yml]: I am trying again, last time as you can see it didn't work...

Comment (ejgallego) [reply .travis.yml]: I should learn quoting rules...

**Comment by SkySkimmer** [comment]
This involves travis quoting rules so the only way to know is to test.
eg on circle defining a variable isn't expanded, so FOO="$BAR" doesn't work.

Gaëtan Gilbert

**Comment by ejgallego** [comment]
Last time I had so much pain that I just dropped the `-unbox-closures` options. Now I've changed parsing so let's see.

**Comment by SkySkimmer** [comment]
#7150 was dropped so this needs rebasing.

**Comment by ejgallego** [comment]
Yup the plan is to implement this a Dune target, so I am keeping this one as a reminded until we advance a bit there.

**Comment by Zimmi48** [comment]
GitLab CI can provide a nice coverage badge BTW for an appropriately tagged job ;-)

**Comment by ejgallego** [comment]
No point on keeping this one open tho, will revive when it is time.
